File: Difficulty: Medium/K Sized Subarray Maximum/k-sized-subarray-maximum.cpp
class Solution {
  public:
    vector<int> maxOfSubarrays(vector<int>& arr, int k) {
        vector<int> ans;
        deque<int> dq;
        int l = 0, r = 0;
        int n = arr.size();

        while (r < n) {

            // step-1 => r ka niptao
            while (!dq.empty() && arr[r] > arr[dq.back()]) {
                dq.pop_back();
            }
            dq.push_back(r);

            // window size control
            if (r - l + 1 > k) {
                if (dq.front() <= l) {
                    dq.pop_front();
                }
                l++;
            }

            // window valid
            if (r - l + 1 == k) {
                ans.push_back(arr[dq.front()]);
            }

            r++;
        }

        return ans;
    }
};

File: Difficulty: Medium/K Sized Subarray Maximum/k-sized-subarray-maximum.js
/**
 * @param {number[]} arr
 * @param {number} k
 * @return {number[]}
 */
class Solution {
    maxOfSubarrays(arr, k) {
        // code here
        let dq = [];
        let l = 0, r = 0;
        let ans = [];
        while(r < arr.length){
            
            while (dq.length > 0 && arr[r] > arr[dq[dq.length - 1]])
                dq.pop();
            dq.push(r);
            
            while(r-l+1 > k){
                if(dq[0] <= l)
                    dq.shift();
                l++;
            }
            
            if(r-l+1 == k){
                ans.push(arr[dq[0]]);
            }
            r++;
        }
        return ans;
    }
}